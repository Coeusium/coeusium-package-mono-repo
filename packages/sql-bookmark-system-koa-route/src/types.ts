export type { ConnectionInfo, Bookmark } from 'sql-bookmark-system';
