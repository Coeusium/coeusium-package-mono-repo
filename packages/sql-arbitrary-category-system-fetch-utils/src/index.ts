export * from './sql-arbitrary-category-system-fetch-utils';
export type { Category, CategorySet } from './types';
