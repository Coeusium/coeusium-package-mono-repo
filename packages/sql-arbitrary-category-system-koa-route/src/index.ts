export * from './sql-arbitrary-category-system-koa-route';
export type { ConnectionInfo, Category, CategorySet } from './types';
