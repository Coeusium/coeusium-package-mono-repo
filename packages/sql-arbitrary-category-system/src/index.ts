export * from './sql-arbitrary-category-system';
export type { ConnectionInfo, Category, CategorySet } from './types';
