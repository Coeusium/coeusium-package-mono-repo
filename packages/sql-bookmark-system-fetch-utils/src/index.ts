export * from './sql-bookmark-system-fetch-utils';
export type { Bookmark } from './types';
