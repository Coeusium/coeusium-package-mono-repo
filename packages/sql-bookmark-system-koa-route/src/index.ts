export * from './sql-bookmark-system-koa-route';
export type { ConnectionInfo, Bookmark } from './types';
