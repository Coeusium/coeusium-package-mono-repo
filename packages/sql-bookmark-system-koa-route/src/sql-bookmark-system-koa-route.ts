import Router from '@koa/router';
import {
  addBookmark,
  deleteBookmark,
  getBookmark,
  listBookmarks,
  bookmarkIdExists,
} from 'sql-bookmark-system';
import compose from 'koa-compose';
import { validateQueryParamsAll } from '@almaclaine/koa-utils';
import { ConnectionInfo } from './types';

export { destroy, setupBookmarkSystem } from 'sql-bookmark-system';

export function createGetBookmarkHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    ctx.body = await getBookmark(dbInfo, id);
  };
}

export function createGetBookmarkListHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { limit, page } = ctx.request.query;
    ctx.body = await listBookmarks(dbInfo, page, limit);
  };
}

export function createBookmarkGetHandler(dbInfo: ConnectionInfo) {
  const getBookmarkHandler = createGetBookmarkHandler(dbInfo);
  const listBookmarkHandler = createGetBookmarkListHandler(dbInfo);
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    if (id) {
      await getBookmarkHandler(ctx, next);
    } else {
      await listBookmarkHandler(ctx, next);
    }
  };
}

export function createPostBookmarkHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { bookmark } = ctx.request.query;
    const id = await addBookmark(dbInfo, bookmark);
    ctx.body = { id };
  };
}

export function createBookmarkDeleteHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    const exists = await bookmarkIdExists(dbInfo, id);
    if (!exists) {
      ctx.status = 412;
      ctx.body = { error: 'No bookmark exists with id ' + id };
      return;
    }

    await deleteBookmark(dbInfo, id);
    ctx.body = {};
  };
}

export function getRouter(dbInfo: ConnectionInfo) {
  const router = new Router();

  const bookmarkGetHandler = createBookmarkGetHandler(dbInfo);
  router.get('/', bookmarkGetHandler);

  const bookmarkDeleteHandler = createBookmarkDeleteHandler(dbInfo);
  router.delete(
    '/',
    compose([validateQueryParamsAll(['id']), bookmarkDeleteHandler]),
  );

  const postBookmarkHandler = createPostBookmarkHandler(dbInfo);
  router.post(
    '/',
    compose([validateQueryParamsAll(['bookmark']), postBookmarkHandler]),
  );

  return router;
}

export function getBookmarkScopedRouter(dbInfo: ConnectionInfo) {
  const router = new Router();
  const routes = getRouter(dbInfo);
  router.use('/bookmark', routes.routes(), routes.allowedMethods());
  return router;
}
