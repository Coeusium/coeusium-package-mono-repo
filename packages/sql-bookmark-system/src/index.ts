export * from './sql-bookmark-system';
export type { ConnectionInfo, Bookmark } from './types';
