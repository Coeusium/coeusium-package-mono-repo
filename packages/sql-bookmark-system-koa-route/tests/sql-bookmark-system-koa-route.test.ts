import * as sqlBookmarkSystem from '../src';

describe('sql-bookmark-system-koa-route-test', () => {
  it('Has Correct API', () => {
    const keys = Object.keys(sqlBookmarkSystem);
    expect(keys.length).toBe(9);
    expect(keys).toMatchSnapshot();
  });
});
