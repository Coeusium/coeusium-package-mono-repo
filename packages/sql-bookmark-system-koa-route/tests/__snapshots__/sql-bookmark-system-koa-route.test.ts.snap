// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql-bookmark-system-koa-route-test Has Correct API 1`] = `
Array [
  "destroy",
  "setupBookmarkSystem",
  "createGetBookmarkHandler",
  "createGetBookmarkListHandler",
  "createBookmarkGetHandler",
  "createPostBookmarkHandler",
  "createBookmarkDeleteHandler",
  "getRouter",
  "getBookmarkScopedRouter",
]
`;
